File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/store/UploadStoreException.java
package eu.cqse.teamscale.jacoco.agent.store;

import okhttp3.ResponseBody;
import retrofit2.Response;

import java.io.IOException;

/**
 * Exception thrown from an upload store. Either during the upload or in the validation process.
 */
public class UploadStoreException extends Exception {

	/** Constructor */
	public UploadStoreException(String message, Exception e) {
		super(message, e);
	}

	/** Constructor */
	public UploadStoreException(String message) {
		super(message);
	}

	/** Constructor */
	public UploadStoreException(String message, Response<ResponseBody> response) {
		super(createResponseMessage(message, response));
	}

	private static String createResponseMessage(String message, Response<ResponseBody> response) {
		try {
			String errorBodyMessage = response.errorBody().string();
			return String.format("%s (%s): \n%s", message, response.code(), errorBodyMessage);
		} catch (IOException | NullPointerException e) {
			return message;
		}
	}
}

File: agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt
package com.teamscale.jacoco.agent.store

import okhttp3.ResponseBody
import retrofit2.Response

import java.io.IOException

/**
 * Exception thrown from an upload store. Either during the upload or in the validation process.
 */
class UploadStoreException(message: String, e: Throwable? = null) : Exception(message, e) {

    companion object {
        fun createForResponseBody(message: String, response: Response<ResponseBody>): Exception {
            return try {
                val errorBodyMessage = response.errorBody()?.string()
                UploadStoreException(String.format("%s (%s): \n%s", message, response.code(), errorBodyMessage))
            } catch (e: IOException) {
                UploadStoreException(message, e)
            }
        }
    }
}
